src/runtime_src/core/pcie/emulation/cpu_em/generic_pcie_hal2/shim.cxx

```diff
@@ -1211,16 +1223,55 @@ namespace xclcpuemhal2 {
       return;
     }
 
+    setDeviceProcessStarted(false);
     std::string socketName = sock->get_name();
-    if(socketName.empty() == false)// device is active if socketName is non-empty
+    // device is active if socketName is non-empty
+    if (!socketName.empty())
     {
 #ifndef _WINDOWS
       xclClose_RPC_CALL(xclClose,this);
 #endif
     }
+   closeMessengerThread();
    saveDeviceProcessOutput();
   }
 
+  void CpuemShim::closeMessengerThread()
+  {
+    if (mMessengerThread.joinable())
+      mMessengerThread.join();
+  }
+
+  void CpuemShim::messagesThread()
+  {
+    auto start_time = std::chrono::high_resolution_clock::now();
+    auto lpath = this->deviceDirectory + "/../../../device_process.log";
```

Comment:
This code for finding the path to the device_process.log is duplicated in multiple locations.   It would be better if a helper function was created to return this value.  This way, if the log ever changes directory, there is only one place that would need to be changed.

Reply:
added a new method